### PR TITLE
JESI-3014

### PR DESCRIPTION
**Pull request by AndreTheHunter**
[JESI-3014](https://jesims.atlassian.net/browse/JESI-3014)
- [x] https://github.com/jesims/backpack/pull/64
- [x] https://github.com/jesims/build-bus/pull/27
- [x] Test in other projects
- [ ] https://github.com/jesims/bindle/pull/1

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,22 @@
-# customs
\ No newline at end of file
+# backpack-test
```
Comment (BadAlgorithm) [inline README.md]: Does this file need to be added or deleted?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+We like to have a theme when testing code that requires data. This projects theme is animal facts (bonus points for Sir David Attenborough quotes).
+[View an example](./test/io/jesi/backpack/collection_test.cljc#L35)
+
+## Helper Utilities
+
+Many additional automation and helper utilities are provided within [backpack.sh](backpack.sh). Invoke `backpack.sh help` to find out more.
+
+## Versioning
+
+> Backpack versions should follow the **Semver** convention.
```
Comment (axrs) [inline CONTRIBUTING.md]: Reference to Backpack

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+Unit tests can then be run by invoking `backpack.sh unit-test` or `backpack.sh unit-test-cljs` respectively.
+
+### Theme: Animal Safari
+
+We like to have a theme when testing code that requires data. This projects theme is animal facts (bonus points for Sir David Attenborough quotes).
+[View an example](./test/io/jesi/backpack/collection_test.cljc#L35)
+
+## Helper Utilities
+
+Many additional automation and helper utilities are provided within [backpack.sh](backpack.sh). Invoke `backpack.sh help` to find out more.
```
Comment (axrs) [inline CONTRIBUTING.md]:
```suggestion
Many additional automation and helper utilities are provided within [customs.sh](customs.sh). Invoke `customs.sh help` to find out more.
```

File: CONTRIBUTING.md (hunk 2 above)
Comment (axrs) [inline CONTRIBUTING.md]:
```suggestion
> Customs versions should follow the **Semver** convention.
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+## Testing
+
+All utilities should be written to support both Clojure and ClojureScript with unit tests written in **CLJC** where practical.
+Unit tests can then be run by invoking `backpack.sh unit-test` or `backpack.sh unit-test-cljs` respectively.
+
+### Theme: Animal Safari
+
+We like to have a theme when testing code that requires data. This projects theme is animal facts (bonus points for Sir David Attenborough quotes).
+[View an example](./test/io/jesi/backpack/collection_test.cljc#L35)
```
Comment (axrs) [inline CONTRIBUTING.md]:
```suggestion
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+**Don't** include more than one feature or fix in a single PR.
+
+**Don't** include changes unrelated to the purpose of the PR.
+
+**Don't** open a new PR if changes are requested. Just push to the same branch and the PR will be updated.
+
+## Testing
+
+All utilities should be written to support both Clojure and ClojureScript with unit tests written in **CLJC** where practical.
+Unit tests can then be run by invoking `backpack.sh unit-test` or `backpack.sh unit-test-cljs` respectively.
```
Comment (axrs) [inline CONTRIBUTING.md]:
```suggestion
Unit tests can then be run by invoking `customs.sh unit-test` or `customs.sh unit-test-cljs` respectively.
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+**DO** ensure the CI checks pass.
+
+**DO** update the [CHANGELOG](CHANGELOG.md) to reflect changes (include your Github identifier).
+
+**DO** update the [CONTRIBUTORS](CONTRIBUTORS.md) to append your name (if not already present).
+
+**DO** notify the project maintainers about any PR that has become stale. We're human and sometimes miss notifications.
+
+**DO** search [crossclj.info](https://crossclj.info/) when considering new additions to see if something already exists.
```
Comment (axrs) [inline CONTRIBUTING.md]:
```suggestion
**DO** search [crossclj.info](https://crossclj.info/) when considering new additions to see if something already exists.
```
Site no longer exists

File: package-dry.json [hunk 7]
```diff
@@ -0,0 +1,37 @@
+{
+  "name": "backpack",
```
Comment (axrs) [inline package-dry.json]:
```suggestion
  "name": "customs",
```

File: package-dry.json [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  "version": "0.0.7",
```
Comment (axrs) [inline package-dry.json]:
```suggestion
  "version": "1.0.0",
```

File: package-dry.json [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+  "dependencyManagement": {
+    "karma": "^4.0.0",
+    "karma-chrome-launcher": "2.2.0",
+    "karma-cljs-test": "0.1.0",
+    "standard": "^12.0.1"
+  },
+  "description": "Clojure(Script) cross-project utilities",
+  "homepage": "https://github.com/jesims/backpack#readme",
```
Comment (axrs) [inline package-dry.json]:
```suggestion
  "homepage": "https://github.com/jesims/customs#readme",
```

File: package-dry.json [hunk 10]
```diff
@@ -0,0 +1,37 @@
+{
+  "name": "backpack",
+  "version": "0.0.7",
+  "dependencyManagement": {
+    "karma": "^4.0.0",
+    "karma-chrome-launcher": "2.2.0",
+    "karma-cljs-test": "0.1.0",
+    "standard": "^12.0.1"
+  },
+  "description": "Clojure(Script) cross-project utilities",
```
Comment (axrs) [inline package-dry.json]:
```suggestion
  "description": "Clojure(Script) cross-project test utilities",
```

File: package-dry.json [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+  "homepage": "https://github.com/jesims/backpack#readme",
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/jesims/backpack.git"
```
Comment (axrs) [inline package-dry.json]:
```suggestion
    "url": "git+https://github.com/jesims/customs.git"
```

File: package-dry.json [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+  },
+  "author": "JESI",
+  "license": "EPL-1.0",
+  "bugs": {
+    "url": "https://github.com/jesims/backpack/issues"
```
Comment (axrs) [inline package-dry.json]:
```suggestion
    "url": "https://github.com/jesims/customs/issues"
```

File: project.clj [hunk 13]
```diff
@@ -0,0 +1,23 @@
+(def VERSION (.trim (slurp "VERSION")))
+
+(defproject io.jesi/customs VERSION
+  :description "Clojure(Script) cross-project testing utilities"
+  :url "https://github.com/jesims/backpack"
```
Comment (axrs) [inline project.clj]:
```suggestion
  :url "https://github.com/jesims/customs"
```

File: src/io/jesi/customs/leiningen.clj [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+  "Runs the `jar` command, returning the path of the built jar"
+  [project]
+  (let [out (let [s (new StringWriter)]
+              (binding [*out* s
+                        *err* s
+                        leiningen.core.main/*info* true
+                        leiningen.core.main/*exit-process?* false]
+                (leiningen.jar/jar project))
+              (-> s str .trim))]
+    (println out)
```
Comment (axrs) [inline src/io/jesi/customs/leiningen.clj]: remove println
```suggestion
```

File: src/io/jesi/customs/leiningen.clj [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+    (some->> out (re-find #"(?<=Created ).+") str .trim)))
+
+;TODO move to backpack?
```
Comment (axrs) [inline src/io/jesi/customs/leiningen.clj]: Does this need to be in BP, or just here? Would prod code need to read a zip?

**Review by axrs — CHANGES_REQUESTED** [review]

File: src/io/jesi/customs/util.cljc [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+  Throws an exception if `timeout` is exceeded.
+  Returns `nil`.")
+  ([f] (wait-for f 1))
+  ([f interval] (wait-for f interval 10000))
+  ([f interval timeout]
+   (if-let [v (f)]
+     #?(:cljs    nil
+        :default v)
+     (let [throw-ex (fn wait-timeout [] (throw (ex-info "Wait timeout" (shorthand timeout f))))]
+       #?(:clj  (let [end-time (+ (System/currentTimeMillis) timeout)]
```
Comment (BadAlgorithm) [inline src/io/jesi/customs/util.cljc:65]: Double space
```suggestion
       #?(:clj (let [end-time (+ (System/currentTimeMillis) timeout)]
```

**Review by BadAlgorithm — CHANGES_REQUESTED** [review]

**Review by BadAlgorithm — DISMISSED** [review]

File: customs.sh [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+
+deploy () {
+	deploy-clojars
+}
+
+deploy-snapshot () {
+	deploy-clojars
+}
+
+script-invoke "$@"
```
Comment (axrs) [inline customs.sh:70]: Loving the use of Bindle to simplify these

File: project.clj [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+                 [org.clojure/clojure :scope "provided"]
+                 [leiningen "2.9.1" :exclusions [org.slf4j/slf4j-nop]]
+                 [org.clojure/tools.namespace "0.3.1"]
+                 ;CLJS
+                 [org.clojure/clojurescript :scope "provided"]
+                 [thheller/shadow-cljs :scope "provided"]
+                 [pjstadig/humane-test-output "0.10.0"]]
+  :profiles {:dev [:parent/dev {:dependencies [[lein-parent "0.3.7"]
+                                               [org.slf4j/slf4j-simple "1.7.30"]]}]})
+
```
Comment (axrs) [inline project.clj]: 👾

File: src/io/jesi/customs/leiningen.clj [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+
+(defn read-project
+  ([] (read-project "project.clj"))
+  ([file] (read-project file [:default]))
+  ([file profiles]
+   (project/ensure-dynamic-classloader)
+   (project/read file profiles)))
+
+(def ^{:arglists '([project])} find-gen-class-ns
+  "Returns a map of ns symbols and it's :gen-class definition (as a map)"
```
Comment (axrs) [inline src/io/jesi/customs/leiningen.clj]:
```suggestion
  "Returns a map of ns symbols and its :gen-class definition (as a map)"
```

File: src/io/jesi/customs/leiningen.clj [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+
+(defn- clojure-ns-paths [project]
+  (apply concat
+    (for [dir (:source-paths project)
+          :let [dir (io/file dir)
+                dir-path (-> dir .toPath)]]
+      (->> dir
+           file-seq
+           (filter (fn [^File f]
+                     (ns-file/file-with-extension? f clojure-file-extensions)))
```
Comment (axrs) [inline src/io/jesi/customs/leiningen.clj]:
```suggestion
                     (bp/partial-right ns-file/file-with-extension? clojure-file-extensions)
```

File: src/io/jesi/customs/leiningen.clj [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+  "Builds the project .jar and checks it only contains the required
+  `.class`, `.clj`, and `META-INF` files.
+  Use with leiningen `:jar-includes` and `:jar-excludes` `project.clj` settings."
+  [project & other-files]
+  (let [expected (sort (concat
+                         (expected-meta-files project)
+                         (find-gen-class-paths project)
+                         (clojure-ns-paths project)
+                         other-files))
+        actual (sort (list-jar project))]
```
Comment (axrs) [inline src/io/jesi/customs/leiningen.clj:154]: Didn't want to do a set comparison?

File: src/io/jesi/customs/strict/kaocha.clj [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+(def- strict-ns (str 'io.jesi.customs.strict))
+
+(defn- remove-ns [form]
+  (sp/transform
+    (sp/walker (bp/and-fn symbol? (bp/compr namespace #{"clojure.core" "cljs.core" strict-ns})))
+    (comp symbol name)
+    form))
+
+(defplugin ::plugin
+
```
Comment (axrs) [inline src/io/jesi/customs/strict/kaocha.clj]: 👾

File: src/io/jesi/customs/strict/kaocha.clj [hunk 23]
```diff
@@ -0,0 +1,27 @@
+(ns io.jesi.customs.strict.kaocha
+  (:refer-clojure :exclude [remove-ns])
+  (:require
+    [com.rpl.specter :as sp]
+    [io.jesi.backpack :as bp]
+    [io.jesi.backpack.macros :refer [def-]]
+    [kaocha.plugin :refer [defplugin]]))
+
+(def- strict-ns (str 'io.jesi.customs.strict))
```
Comment (axrs) [inline src/io/jesi/customs/strict/kaocha.clj:9]:
```suggestion
(def- strict-ns "io.jesi.customs.strict")
```
Less characters. Does the quoted symbol give us any other benefit?

File: test/io/jesi/customs/leiningen_test.clj [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+
+    (testing "deletes target/classes dir when done"
+      (let [jar-path (lein/build-jar project-with-install)]
+        (is (seq jar-path))
+        (is (-> jar-path io/file (.exists)))
+        (is (not (-> project-with-install :target-path (str "/classes") io/file (.exists))))))))
+
+(deftest is-slim-jar-test
+
+  (lein/is-slim-jar project-with-install "FILE"))
```
Comment (axrs) [inline test/io/jesi/customs/leiningen_test.clj]:
```suggestion

  (testing "is-slim-jar"
    (lein/is-slim-jar project-with-install "FILE")))
```

File: test/io/jesi/customs/strict/kaocha_test.clj [hunk 25]
```diff
@@ -0,0 +1,7 @@
+(ns io.jesi.customs.strict.kaocha-test
+  (:refer-clojure :exclude [=])
+  (:require
+    [io.jesi.customs.strict :refer :all]))
+
+(deftest ^:kaocha/skip kaocha-diff-test
+  (is= {:a 1} {:a 2}))
```
Comment (axrs) [inline test/io/jesi/customs/strict/kaocha_test.clj:9]: This is confusing. Is this just to check it works? Should we add a comment as to why it's skipped or actually `not=`

**Review by axrs — CHANGES_REQUESTED** [review]

File: project.clj (hunk 18 above)
Comment (AndreTheHunter) [reply project.clj]:
```suggestion
```

File: src/io/jesi/customs/leiningen.clj (hunk 21 above)
Comment (AndreTheHunter) [reply src/io/jesi/customs/leiningen.clj:154]: Sorting makes the diff a bit nicer

File: src/io/jesi/customs/strict/kaocha.clj (hunk 22 above)
Comment (AndreTheHunter) [reply src/io/jesi/customs/strict/kaocha.clj]:
```suggestion
```

File: src/io/jesi/customs/strict/kaocha.clj (hunk 23 above)
Comment (AndreTheHunter) [reply src/io/jesi/customs/strict/kaocha.clj:9]: Having it quoted means that Cursive can link to the ns (cmd+b). Helps prevent typos

File: test/io/jesi/customs/strict/kaocha_test.clj (hunk 25 above)
Comment (AndreTheHunter) [reply test/io/jesi/customs/strict/kaocha_test.clj:9]: Yeah I couldn't figure out how to to intercept kaocha reporting so had to have a failing test to test the diff works. I'll add a commment

File: src/io/jesi/customs/leiningen.clj (hunk 21 above)
Comment (axrs) [reply src/io/jesi/customs/leiningen.clj:154]: 👍

**Review by axrs — DISMISSED** [review]

**Review by BadAlgorithm — DISMISSED** [review]

File: project.clj [hunk 26]
```diff
@@ -0,0 +1,26 @@
...
+  :dependencies [[io.jesi/backpack "4.2.1"]                 ;TODO use managed version (at least 5.0)
+                 ;CLJ
+                 [org.clojure/clojure :scope "provided"]
+                 [leiningen "2.9.1" :exclusions [org.slf4j/slf4j-nop]]
+                 [org.clojure/tools.namespace "0.3.1"]
+                 ;CLJS
+                 [org.clojure/clojurescript :scope "provided"]
+                 [thheller/shadow-cljs :scope "provided"]
+                 [pjstadig/humane-test-output "0.10.0"]]
+  :profiles {:dev [:parent/dev {:dependencies [[lein-parent "0.3.7"]
```
Comment (BadAlgorithm) [inline project.clj]: Does lein-parent need to be a dep (I thought it was only needed to be listed as plugin)?

**Review by BadAlgorithm — CHANGES_REQUESTED** [review]

File: project.clj (hunk 26 above)
Comment (AndreTheHunter) [reply project.clj]: No it does not. Removed

File: customs.sh [hunk 27]
```diff
@@ -0,0 +1,71 @@
...
+
+## clean:
+clean () {
+	lein-clean
+}
+
+## test:
+## args: [-r]
+test () {
+	-test-clj "$@"
```
Comment (axrs) [inline customs.sh:36]: Surely this fails shellcheck 2215?

File: src/io/jesi/customs/leiningen.clj [hunk 28]
```diff
@@ -0,0 +1,153 @@
...
+
+(def ^{:arglists '([project])} find-gen-class-ns
+  "Returns a map of ns symbols and its :gen-class definition (as a map)"
+  (memoize
+    (fn find-gen-class-ns-fn [project]
+      (->> (for [dir (:source-paths project)]
+             (->> dir
+                  io/file
+                  ns-find/find-ns-decls-in-dir
+                  (map (fn [ns-decl]
```
Comment (axrs) [inline src/io/jesi/customs/leiningen.clj]:
```suggestion
                  (keep (fn [ns-decl]
```
Should filter `nil`s for you

File: src/io/jesi/customs/leiningen.clj [hunk 29]
```diff
@@ -0,0 +1,153 @@
...
+  ([file] (read-project file [:default]))
+  ([file profiles]
+   (project/ensure-dynamic-classloader)
+   (project/read file profiles)))
+
+(def ^{:arglists '([project])} find-gen-class-ns
+  "Returns a map of ns symbols and its :gen-class definition (as a map)"
+  (memoize
+    (fn find-gen-class-ns-fn [project]
+      (->> (for [dir (:source-paths project)]
```
Comment (axrs) [inline src/io/jesi/customs/leiningen.clj]: Why `for` and not `reduce`?

**Review by axrs — CHANGES_REQUESTED** [review]

File: customs.sh (hunk 27 above)
Comment (AndreTheHunter) [reply customs.sh:36]: I guess it's happy with it since it takes args

**Review by axrs — APPROVED** [review]

**Review by BadAlgorithm — APPROVED** [review]

**Pull request merged**